### PR TITLE
Alarm update

### PR DESCRIPTION
**Pull request by jarongrigat**
I pushed as you, but whatever. These are the fixes from apollo plus an update to the AlarmMonitor, which should be working as intended now.

### CODE REVIEW

File: Doberman/AlarmMonitor.py [hunk 1]
```diff
@@ -174,20 +174,23 @@ def check_for_alarms(self):
...
-                'logged': {'$push': {'$toDate': '$_id'}},
-                'msgs': {'$push': '$msg'}}}]):
-            message = '\n'.join([f'{d.iso_format()}: {m}' for d,m in zip(doc['logged'], doc['msgs'])])
-            self.send_message(doc['_id'], message)
+        alarms = {}
+        for doc in self.db.read_from_db(db_col, doc_filter):
+            level = doc['level'] + doc['escalation']
+            logged = datetime.fromtimestamp(int(str(doc['_id'])[:8], 16))
+            if level not in alarms:
+                alarms[level] = {'logged': [], 'msgs': []}
```
Comment (darrylmasson) [inline Doberman/AlarmMonitor.py:185]: I'm pretty sure this fails. You can't use a simple int as a dictionary key, but if you tuple it `(level,)` then it works.

File: Doberman/Node.py [hunk 2]
```diff
@@ -236,8 +246,12 @@ def merge_field(self, field, packages):
     def process(self, packages):
         new_package = {}
         common_keys = set(packages[0].keys())
+        uncommon_keys = set(packages[0].keys())
+        self.logger.debug(f'Package: {packages[0]}')
         for p in packages[1:]:
+            self.logger.debug(f'Package: {p}')
```
Comment (darrylmasson) [inline Doberman/Node.py:252]: I think we can do away with these two log statements as well, now that we fixed the problem they were added to help trace.

File: Doberman/Database.py [hunk 3]
```diff
@@ -243,20 +225,43 @@ def set_pipeline_value(self, name, kvp):
...
+        """
+        recipient_groups = self.get_experiment_config('alarm', 'recipients')
+        if len(recipient_groups) < level:
+            recipient_groups = ['everyone']
+        else:
+            recipient_groups = recipient_groups[level]
+        recipient_names = set()
+        for group in recipient_groups:
+            if group == 'shifters':
+                for doc in self.read_from_db('contacts', {'on_shift': 'true'}):
```
Comment (darrylmasson) [inline Doberman/Database.py:255]: Are we storing `on_shift` as a string or a bool?

File: Doberman/Database.py [hunk 4]
```diff
@@ -243,20 +225,43 @@ def set_pipeline_value(self, name, kvp):
...
+        return protocols[level]
+
+    def get_message_recipients(self, level):
+        """
+        Gets message recipients for the specified alarm level
+
+        :param level: which alarm level is in question (0, 1, etc)
+        :returns: set of recipient names
+        """
+        recipient_groups = self.get_experiment_config('alarm', 'recipients')
```
Comment (darrylmasson) [inline Doberman/Database.py:247]: Don't forget to update the docs with info about this field

**Review by darrylmasson — CHANGES_REQUESTED** [review]
A couple of minor tweaks, but otherwise it looks solid

File: Doberman/AlarmMonitor.py (hunk 1 above)
Comment (jarongrigat) [reply Doberman/AlarmMonitor.py:183]: It does work.

File: Doberman/Node.py (hunk 2 above)
Comment (jarongrigat) [reply Doberman/Node.py]: Done

File: Doberman/Database.py (hunk 3 above)
Comment (jarongrigat) [reply Doberman/Database.py]: I changed it to bool

File: Doberman/Database.py (hunk 4 above)
Comment (jarongrigat) [reply Doberman/Database.py:247]: Done

File: Doberman/AlarmMonitor.py (hunk 1 above)
Comment (darrylmasson) [reply Doberman/AlarmMonitor.py:183]: I feel like this got patched in a recent python version, I recall this failing in the past. As long as it works now (which it does seem to do)

**Review by darrylmasson — APPROVED** [review]

**Pull request merged**